tools/releaser/main.go

```diff
@@ -0,0 +1,289 @@
+	if verbose {
+		fmt.Println("Writing new deps.bzl")
+	}
+	if _, err := newDepsBuffer.Write(depsMaybeBuff.Bytes()); err != nil {
+		return err
+	}
+	if err := os.WriteFile("deps.bzl", newDepsBuffer.Bytes(), 0644); err != nil {
+		return err
+	}
+
+	// update WORKSPACE
+	if verbose {
+		fmt.Println("Writing new WORKSPACE")
+	}
+	directivelessWorkspace := workspaceWithoutDirectives.Bytes()
+	newWorkspace := make([]byte, 0, len(directivelessWorkspace))
+	newWorkspace = append(newWorkspace, directivelessWorkspace[:directiveStart]...)
+	newWorkspace = append(newWorkspace, workspaceDirectivesBuff.Bytes()...)
+	newWorkspace = append(newWorkspace, directivelessWorkspace[directiveStart:]...)
+
+	if err := os.WriteFile("WORKSPACE", newWorkspace, 0644); err != nil {
+		return err
+	}
+
+	// cleanup before final gazelle run
+	if verbose {
+		fmt.Println("Cleaning up temporary files")
+	}
+	if err := os.Remove(_tmpBzl); err != nil {
+		return err
+	}
+
+	if verbose {
+		fmt.Println("Running final gazelle run, and copying some language specific build files.")
+	}
+	finalizationCommands := []struct {
+		cmd  string
+		args []string
+	}{
+		{cmd: "bazel", args: []string{"run", "//:gazelle"}},
+		{cmd: "bazel", args: []string{"build", "//language/proto:known_imports"}},
+		{cmd: "cp", args: []string{"-f", "bazel-bin/language/proto/known_imports.go", "language/proto/known_imports.go"}},
+		{cmd: "bazel", args: []string{"build", "//language/proto:known_proto_imports"}},
+		{cmd: "cp", args: []string{"-f", "bazel-bin/language/proto/known_proto_imports.go", "language/proto/known_proto_imports.go"}},
+		{cmd: "bazel", args: []string{"build", "//language/proto:known_go_imports"}},
+		{cmd: "cp", args: []string{"-f", "bazel-bin/language/proto/known_go_imports.go", "language/proto/known_go_imports.go"}},
+	}
+	for _, c := range finalizationCommands {
+		if out, err := exec.CommandContext(ctx, c.cmd, c.args...).CombinedOutput(); err != nil {
+			fmt.Println(string(out))
+			return err
+		}
+	}
+	return nil
+}
+
+func getWorkspaceWithouthDirectives(workspaceBuffer, workspaceWithoutDirectives *bytes.Buffer) (int, error) {
+	var directiveStart int
+	var directiveFound bool
+	workspaceScanner := bufio.NewScanner(workspaceBuffer)
+	for workspaceScanner.Scan() {
+		if bytes.HasPrefix(workspaceScanner.Bytes(), []byte("# gazelle:repository go_repository")) {
+			directiveFound = true
+			continue
+		}
+		n, err := workspaceWithoutDirectives.Write(workspaceScanner.Bytes())
+		if err != nil {
+			return directiveStart, err
+		}
+		_, err = workspaceWithoutDirectives.Write([]byte("\n"))
+		if err != nil {
+			return directiveStart, err
+		}
+		if !directiveFound {
+			directiveStart += n + 1
+		}
+	}
+
+	return directiveStart, nil
+}
+
+func getBuffsFromTmp(depsMaybeBuff, workspaceDirectivesBuff *bytes.Buffer) error {
+	tmpbzl, err := os.ReadFile(_tmpBzl)
+	if err != nil {
+		return err
+	}
+
+	attributeRegex := regexp.MustCompile(`^\s+(name|importpath) = "(.+)",$`)
```

Comment:
You can call Gazelle's [LoadMacroFile](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle/rule#LoadMacroFile) function to parse the whole `tmpbzl` and iterate through the rules and build the rule data structure. At the end, you can call [Format](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle/rule#File.Format) to convert it to bytes.

Reply:
I originally went down this road but could not figure out how to modify directives using the returned rule.File since the documentation for `Directives` specifies it should not be modified after reading the file. Let me know if I missed an obvious alternative to manipulating this set.

Reply:
This function has 3 parts:

1. Parsing the tmpbzl, for which you can use Gazelle's [LoadMacroFile](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle/rule#LoadMacroFile) function
2. Building depsMaybe. You can first create an [EmptyFile](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle/rule#EmptyFile), build `_maybe` rules, [insert](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle/rule#Rule.Insert) them into the File and [Save](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle/rule#File.Save) them to disk or [Format](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle/rule#File.Format) them into `[]byte`.
3. Building `# gazelle:repository` directives. This part is simple enough that I think it's fine with the current approach.